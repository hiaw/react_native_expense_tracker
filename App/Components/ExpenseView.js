import React, { Component } from 'react'
import { Alert, View, Button } from 'react-native'
import { observable } from 'mobx'
import { observer } from 'mobx-react'
import Moment from 'moment'
import { FormLabel, FormInput } from 'react-native-elements'
import * as firebase from 'firebase'

import store from '../Model/MainStore.js'
import styles from './Styles/ExpenseView.style.js'

@observer
export default class ExpenseView extends Component {
  @observable date = ''
  @observable description = ''
  @observable amount = ''
  @observable comment = ''
  
  constructor (props) {
    super(props)

    this.date = props.expense.date
    this.description = props.expense.description
    this.amount = props.expense.amount
    this.comment = props.expense.comment

    let userMobilePath = '/user/' + store.userDevice.userId +
       '/expenses/' + props.expense.key
    this.expenseRef = firebase.database().ref(userMobilePath)
  }

  save () {
    this.expenseRef.set({
      date: Moment(this.date).valueOf(),
      description: this.description,
      amount: this.amount,
      comment: this.comment
    })
  }

  delete () {
    Alert.alert(
      'Delete this expense',
      'Are you sure?',
      [
        {text: 'Yes', onPress: () => { this.expenseRef.remove() }},
        {text: 'Cancel', onPress: () => {}, style: 'cancel'}
      ]
    )
  }

  changeDate(date) {
    console.log(date)
  }

  changeTime(time) {
    console.log(time)
  }

  render () {
    let dateText = Moment(this.date).format('DD/MM/YY')
    let timeText = Moment(this.date).format('HH:mm')

    return (
      <View style={styles.container}>
        <FormLabel>Date</FormLabel>
        <FormInput value={dateText}
          onChangeText={(t) => {this.changeDate(t)}}/>

        <FormLabel>Time</FormLabel>
        <FormInput value={timeText}
          onChangeText={(t) => {this.changeTime(t)}}/>

        <FormLabel>Amount</FormLabel>
        <FormInput value={this.amount}
          onChangeText={(t) => {this.amount = t}}/>

        <FormLabel>Description</FormLabel>
        <FormInput value={this.description}
          onChangeText={(t) => {this.description= t}}/>

        <FormLabel>Comment</FormLabel>
        <FormInput value={this.comment}
          onChangeText={(t) => {this.comment= t}}/>

        <Button title='Save'
          onPress={() => this.save()} />
        <Button title='Delete'
          color='red'
          onPress={() => this.delete()} />
      </View>
    )
  }
}
