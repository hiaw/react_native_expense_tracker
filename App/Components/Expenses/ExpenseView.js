import React, { Component } from 'react'
import { Alert, View, Button, DatePickerIOS } from 'react-native'
import { observable } from 'mobx'
import { observer } from 'mobx-react'
import Moment from 'moment'
import { FormLabel, FormInput } from 'react-native-elements'
import { Actions } from 'react-native-router-flux'

import store from '../../Model/MainStore.js'
import styles from './ExpenseView.style.js'

@observer
export default class ExpenseView extends Component {
  @observable editing = false
  @observable date = new Date()
  @observable description = ''
  @observable amount = ''
  @observable comment = ''

  constructor (props) {
    super(props)

    this.expenseService = props.app.service('expenses')

    if (props.expense) {
      this.editing = true
      this.date = Moment(props.expense.date).toDate()
      this.description = props.expense.description
      this.amount = props.expense.amount
      this.comment = props.expense.comment
    }
  }

  add() {
    this.expenseService.create({
      owner: store.userDevice.userId,
      date: this.date,
      description: this.description,
      amount: Number(this.amount),
      comment: this.comment
    }).then((res) => {
      Actions.pop()
    })
  }

  save () {
    this.expenseService.update(this.props.expense._id, {
      owner: this.props.expense.owner,
      date: this.date,
      description: this.description,
      amount: Number(this.amount),
      comment: this.comment
    }).then((res) => {
      Actions.pop()
    })
  }

  delete () {
    Alert.alert(
      'Delete this expense',
      'Are you sure?',
      [
        {text: 'Yes', onPress: () => {
          this.expenseService.remove(this.props.expense._id)
            .then((res) => {
              Actions.pop()
            })
        }},
        {text: 'Cancel', onPress: () => {}, style: 'cancel'}
      ]
    )
  }

  render () {
    let dateText = Moment(this.date).format('DD/MM/YY')
    let timeText = Moment(this.date).format('HH:mm')

    return (
      <View style={styles.container}>
        <FormLabel>Date</FormLabel>
        <DatePickerIOS
          date={this.date}
          mode="datetime"
          onDateChange={(t) => { this.date = t }}
        />

        <FormLabel>Amount</FormLabel>
        <FormInput value={this.amount.toString()}
          keyboardType='numeric'
          onChangeText={(t) => {this.amount = t}}/>

        <FormLabel>Description</FormLabel>
        <FormInput value={this.description}
          onChangeText={(t) => {this.description= t}}/>

        <FormLabel>Comment</FormLabel>
        <FormInput value={this.comment}
          inputStyle={{height: 100}}
          multiline={true}
          numberOfLines={4}
          onChangeText={(t) => {this.comment= t}}/>

        { this.editing? this.renderEditButtons(): this.renderAddButton()}

      </View>
    )
  }

  renderAddButton() {
    return (
      <Button title='Add'
        onPress={() => this.add()} />
    )
  }

  renderEditButtons() {
    return (
      <View>
        <Button title='Save'
          onPress={() => this.save()} />
        <Button title='Delete'
          color='red'
          onPress={() => this.delete()} />
      </View>
    )
  }
}
