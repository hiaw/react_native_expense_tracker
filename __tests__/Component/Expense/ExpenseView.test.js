import 'react-native'
import React from 'react'
import ExpenseView from '../../../App/Components/Expenses/ExpenseView.js'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

let expense = {
  _id: 'adsflkja',
  date: new Date(),
  description: 'description',
  amount: 99.00,
  comment: 'Some long comment'
}

let app = { service: () => {}}

it('renders correctly', () => {
  const tree = renderer.create(
    <ExpenseView expense={expense} app={app} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
