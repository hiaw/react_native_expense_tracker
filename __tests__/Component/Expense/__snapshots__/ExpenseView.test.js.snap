// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 65,
    }
  }
>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 15,
              "fontWeight": "bold",
              "marginBottom": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 15,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Date
    </Text>
  </View>
  <View
    style={undefined}
  >
    <RCTDatePicker
      date={949316400000}
      maximumDate={undefined}
      minimumDate={undefined}
      minuteInterval={undefined}
      mode="datetime"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 216,
        }
      }
      timeZoneOffsetInMinutes={undefined}
    />
  </View>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 15,
              "fontWeight": "bold",
              "marginBottom": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 15,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Amount
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#bdc6cf",
          "borderBottomWidth": 1,
          "marginLeft": 20,
          "marginRight": 20,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize={undefined}
      autoCorrect={undefined}
      autoFocus={undefined}
      blurOnSubmit={undefined}
      clearButtonMode={undefined}
      clearTextOnFocus={undefined}
      dataDetectorTypes={undefined}
      defaultValue={undefined}
      editable={undefined}
      enablesReturnKeyAutomatically={undefined}
      inlineImageLeft={undefined}
      inlineImagePadding={undefined}
      keyboardAppearance={undefined}
      keyboardType="numeric"
      maxLength={undefined}
      multiline={undefined}
      numberOfLines={undefined}
      onBlur={undefined}
      onChange={undefined}
      onChangeText={[Function]}
      onContentSizeChange={undefined}
      onEndEditing={undefined}
      onFocus={undefined}
      onKeyPress={undefined}
      onLayout={undefined}
      onSelectionChange={undefined}
      onSubmitEditing={undefined}
      placeholder={undefined}
      placeholderTextColor={undefined}
      returnKeyLabel={undefined}
      returnKeyType={undefined}
      secureTextEntry={undefined}
      selectTextOnFocus={undefined}
      selectionColor="#86939e"
      selectionState={undefined}
      style={
        Array [
          Object {
            "color": "#86939e",
            "fontSize": 17.5,
            "height": 36,
          },
          undefined,
        ]
      }
      underlineColorAndroid={undefined}
      value="99"
    />
  </View>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 15,
              "fontWeight": "bold",
              "marginBottom": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 15,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Description
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#bdc6cf",
          "borderBottomWidth": 1,
          "marginLeft": 20,
          "marginRight": 20,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize={undefined}
      autoCorrect={undefined}
      autoFocus={undefined}
      blurOnSubmit={undefined}
      clearButtonMode={undefined}
      clearTextOnFocus={undefined}
      dataDetectorTypes={undefined}
      defaultValue={undefined}
      editable={undefined}
      enablesReturnKeyAutomatically={undefined}
      inlineImageLeft={undefined}
      inlineImagePadding={undefined}
      keyboardAppearance={undefined}
      keyboardType={undefined}
      maxLength={undefined}
      multiline={undefined}
      numberOfLines={undefined}
      onBlur={undefined}
      onChange={undefined}
      onChangeText={[Function]}
      onContentSizeChange={undefined}
      onEndEditing={undefined}
      onFocus={undefined}
      onKeyPress={undefined}
      onLayout={undefined}
      onSelectionChange={undefined}
      onSubmitEditing={undefined}
      placeholder={undefined}
      placeholderTextColor={undefined}
      returnKeyLabel={undefined}
      returnKeyType={undefined}
      secureTextEntry={undefined}
      selectTextOnFocus={undefined}
      selectionColor="#86939e"
      selectionState={undefined}
      style={
        Array [
          Object {
            "color": "#86939e",
            "fontSize": 17.5,
            "height": 36,
          },
          undefined,
        ]
      }
      underlineColorAndroid={undefined}
      value="description"
    />
  </View>
  <View
    style={
      Array [
        Object {},
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": undefined,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 15,
              "fontWeight": "bold",
              "marginBottom": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "marginTop": 15,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Comment
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderBottomColor": "#bdc6cf",
          "borderBottomWidth": 1,
          "marginLeft": 20,
          "marginRight": 20,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize={undefined}
      autoCorrect={undefined}
      autoFocus={undefined}
      blurOnSubmit={undefined}
      clearButtonMode={undefined}
      clearTextOnFocus={undefined}
      dataDetectorTypes={undefined}
      defaultValue={undefined}
      editable={undefined}
      enablesReturnKeyAutomatically={undefined}
      inlineImageLeft={undefined}
      inlineImagePadding={undefined}
      keyboardAppearance={undefined}
      keyboardType={undefined}
      maxLength={undefined}
      multiline={true}
      numberOfLines={4}
      onBlur={undefined}
      onChange={undefined}
      onChangeText={[Function]}
      onContentSizeChange={undefined}
      onEndEditing={undefined}
      onFocus={undefined}
      onKeyPress={undefined}
      onLayout={undefined}
      onSelectionChange={undefined}
      onSubmitEditing={undefined}
      placeholder={undefined}
      placeholderTextColor={undefined}
      returnKeyLabel={undefined}
      returnKeyType={undefined}
      secureTextEntry={undefined}
      selectTextOnFocus={undefined}
      selectionColor="#86939e"
      selectionState={undefined}
      style={
        Array [
          Object {
            "color": "#86939e",
            "fontSize": 17.5,
            "height": 36,
          },
          Object {
            "height": 100,
          },
        ]
      }
      underlineColorAndroid={undefined}
      value="Some long comment"
    />
  </View>
  <View>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#0C42FD",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Save
        </Text>
      </View>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#0C42FD",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "red",
              },
            ]
          }
        >
          Delete
        </Text>
      </View>
    </View>
  </View>
</View>
`;
